pkg/urbit/jets/e/en_json.c

```diff
@@ -0,0 +1,438 @@
+          break;
+        }
+
+        case 92:
+        {
+          buf_u = _append_text(buf_u, _JSON_BAS, 2);
+          break;
+        }
+
+        case 127:
+        {
+          buf_u = _append_text(buf_u, _JSON_DEL, 6);
+          break;
+        }
+
+        default:
+        {
+          buf_u = _append_char(buf_u, c_y);
+        }
+      }
+    }
+    buf_u = _append_char(buf_u, '"');
+
+    return buf_u;
+  }
+
+  json_buffer*
+  _serialize_array(json_buffer *buf_u, u3_noun a)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+
+    buf_u = _append_char(buf_u, '[');
+
+    if ( u3_nul != a ) {
+      u3_noun i, t = a;
+
+      while ( u3_nul != t ) {
+        u3x_cell(t, &i, &t);
+        buf_u = _serialize(buf_u, i);
+        buf_u = _append_char(buf_u, ',');
+      }
+
+      if ( buf_u == NULL )
+        return NULL;
+      
+      --buf_u->len_w;
+    }
+
+    buf_u = _append_char(buf_u, ']');
+    return buf_u;
+  }
+
+  json_buffer*
+  _serialize_object_helper(json_buffer *buf_u, u3_noun a)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+    else if ( u3_nul == a ) {
+      return buf_u;
+    }
+
+    u3_noun n_a, l_a, r_a;
+    u3_noun pn_a, qn_a;
+    u3x_trel(a, &n_a, &l_a, &r_a);
+    u3x_cell(n_a, &pn_a, &qn_a);
+
+    // order is important to match unjetted tree traversal
+    //
+    buf_u = _serialize_object_helper(buf_u, r_a);
+
+    buf_u = _serialize_string(buf_u, pn_a);
+    buf_u = _append_char(buf_u, ':');
+    buf_u = _serialize(buf_u, qn_a);
+    buf_u = _append_char(buf_u, ',');
+
+    buf_u = _serialize_object_helper(buf_u, l_a);
+
+    return buf_u;
+  }
+
+  json_buffer*
+  _serialize_object(json_buffer *buf_u, u3_noun a)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+
+    buf_u = _append_char(buf_u, '{');
+
+    if ( u3_nul != a ) {
+      buf_u = _serialize_object_helper(buf_u, a);
+      if ( buf_u == NULL ) {
+        return NULL;
+      }
+      --buf_u->len_w;
+    }
+
+    buf_u = _append_char(buf_u, '}');
+    return buf_u;
+  }
+
+  json_buffer*
+  _serialize(json_buffer *buf_u, u3_noun a)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+
+    if ( u3_nul == a ) {
+      return _serialize_null(buf_u);
+    }
+    
+    u3_noun s, p;
+    u3x_cell(a, &s, &p);
+    if ( _(u3du(s)) || _(u3a_is_dog(s)) ) {
+      return _encode_fail(buf_u);
+    }
+
+    switch ( s ) {
+      case c3__a:
+      {
+        buf_u = _serialize_array(buf_u, p);
+        break;
+      }
+
+      case c3__o:
+      {
+        buf_u = _serialize_object(buf_u, p);
+        break;
+      }
+
+      case c3__b:
+      {
+        buf_u = _serialize_loobean(buf_u, p);
+        break;
+      }
+
+      case c3__n:
+      {
+        buf_u = _serialize_number(buf_u, p);
+        break;
+      }
+
+      case c3__s:
+      {
+        buf_u = _serialize_string(buf_u, p);
+        break;
+      }
+
+      default:
+      {
+        return _encode_fail(buf_u);
+      }
```

Comment:
As we discussed, jetting hoon `?-` expressions with switch statements is fairly subtle (and I had it wrong ..). The choices are to either make the last pattern/case fall through to the default case, or to bail non-deterministically in the default case. After further review, I prefer the latter -- it better captures the parts of the current behavior we care about, and is more flexible going forward (in case we decide to change the hoon compiler's behavior).

So, concretely, the default case here should `u3m_bail(c3__fail)`.

Reply:
Done